### ISSUE TITLE
non-strings are not piped to external commands

### ISSUE THREAD

**Issue by ineu** [nushell/nushell#14452]
### Describe the bug

If a function returns value other than string, it is silently ignored when fed to an external command via pipe. This is unexpected.

### How to reproduce

/home/ineu ▶ '123' | cat
123
/home/ineu ▶ 123 | cat
/home/ineu ▶ 123 | into string | cat
123

### Expected behavior

If a function returns a value that is not a string (like int, float, bool etc) it should be implicitly converted into string when fed to an external command

### Configuration

| key                | value                                                            |
| ------------------ | ---------------------------------------------------------------- |
| version            | 0.100.0                                                          |
| major              | 0                                                                |
| minor              | 100                                                              |
| patch              | 0                                                                |
| branch             | makepkg                                                          |
| commit_hash        | 30f98f7e64a63a159ce70f02aa6c4c1196cd9ca7                         |
| build_os           | linux-x86_64                                                     |
| build_target       | x86_64-unknown-linux-gnu                                         |
| rust_version       | rustc 1.82.0 (f6e511eec 2024-10-15) (Arch Linux rust 1:1.82.0-2) |
| cargo_version      | cargo 1.82.0 (8f40fc59f 2024-08-21)                              |
| build_time         | 2024-11-13 08:00:49 +00:00                                       |
| build_rust_channel | release                                                          |
| allocator          | mimalloc                                                         |
| features           | default, sqlite, trash                                           |
| installed_plugins  | example 0.99.0, formats 0.99.1                                   |

**Comment by fdncred** [comment]
why? what's the use case for this?

**Comment by sholderbach** [comment]
I don't think we should implicitly coerce but it should raise a mismatch if its not string or binary data going to an external.

**Comment by ineu** [comment]
> why? what's the use case for this?

I originally faced this bug while creating a script to manage screen brightness. The logic to increase the brightness is:
1. Read current and max absolute (device-specific) values from /sys
2. Calculate the new value (which obviously emits an integer)
3. Write the new value back to /sys

So my code was:
```nu
# Increase backlight
def "main up" [percent = 10] {
    let new = [((absolute_current | absolute_to_percent) + $percent), 100] | math min

    $new | percent_to_absolute | sudo tee $current_file
}
```
and it took me some time to realize why `sudo tee` creates an empty file.

> I don't think we should implicitly coerce but it should raise a mismatch if its not string or binary data going to an external.

That's a great option. Much better than silently ignoring the output.

**Comment by IanManske** [comment]
We already implicitly convert for arguments to external commands, I don't see why we can't for the piped input as well. As long as the conversion is consistent and documented, I think that would be a little more ergonomic (e.g., #14090). 

Alternatively, we could remove the implicit conversion for arguments if we want stricter typing. Either way, the two should match.

**Comment by 132ikl** [comment]
Related: #14452

**Comment by 132ikl** [comment]
I'm thinking we should explicitly list pipeline input types for `run-external` (unfortunately, this will be a shell error instead of a parse error until we fix the [conditionally discarded pipeline input problem](https://hackmd.io/@132ikl/HJTwKCoq1l)) only convert primitives with an obvious conversion.

I agree with @IanManske on the point of argument conversions matching pipeline input conversions. I think we should only convert primitives which have an obvious conversion, but not things like dates, lists, records, etc. I think a decent heuristic is that the string representation is the same as the literal typed into the REPL then we implicitly convert, otherwise we don't.
